File: exercises/static/exercises/laser_mapping/python_template/ros2_humble/map.py
import numpy as np
import math
import random
from math import pi as pi
import cv2


class Map:
    def __init__(self, pose_getter, noisy_pose_getter):
        self.pose_getter = pose_getter
        self.noisy_pose_getter = noisy_pose_getter

    def RTx(self, angle, tx, ty, tz):
        RT = np.matrix(
            [
                [1, 0, 0, tx],
                [0, math.cos(angle), -math.sin(angle), ty],
                [0, math.sin(angle), math.cos(angle), tz],
                [0, 0, 0, 1],
            ]
        )
        return RT

    def RTy(self, angle, tx, ty, tz):
        RT = np.matrix(
            [
                [math.cos(angle), 0, math.sin(angle), tx],
                [0, 1, 0, ty],
                [-math.sin(angle), 0, math.cos(angle), tz],
                [0, 0, 0, 1],
            ]
        )
        return RT

    def RTz(self, angle, tx, ty, tz):
        RT = np.matrix(
            [
                [math.cos(angle), -math.sin(angle), 0, tx],
                [math.sin(angle), math.cos(angle), 0, ty],
                [0, 0, 1, tz],
                [0, 0, 0, 1],
            ]
        )
        return RT

    def RTVacuum(self):
        RTz = self.RTz(pi / 2, 50, 70, 0)
        return RTz

    def getRobotCoordinates(self):
        pose = self.pose_getter()
        x = pose.x
        y = pose.y

        scale_y = -23.53
        offset_y = -31.95
        y = scale_y * (offset_y - y)

        scale_x =  -23.58
        offset_x = -20.36
        x = scale_x * (offset_x - x)

        return x, y, pose.yaw

    def getRobotCoordinatesWithNoise(self):
        pose = self.noisy_pose_getter()
        x = pose.x
        y = pose.y

        scale_y = -23.53
        offset_y = -31.95
        y = scale_y * (offset_y - y)

        scale_x =  -23.58
        offset_x = -20.36
        x = scale_x * (offset_x - x)

        return x, y, pose.yaw

    # Function to reset
    def reset(self):
        # Nothing to do, service takes care!
        pass

File: exercises/laser_mapping/cpp_lib/src/Map.cpp
#include "Map.hpp"
#include <cmath>

Map::Map(std::function<Pose3d()> pose_getter, std::function<Pose3d()> noisy_pose_getter)
    : pose_getter_(pose_getter), noisy_pose_getter_(noisy_pose_getter)
{
}

cv::Mat Map::rtx(double angle, double tx, double ty, double tz)
{
    return (cv::Mat_<double>(4, 4) << 
        1, 0, 0, tx,
        0, std::cos(angle), -std::sin(angle), ty,
        0, std::sin(angle), std::cos(angle), tz,
        0, 0, 0, 1);
}

cv::Mat Map::rty(double angle, double tx, double ty, double tz)
{
    return (cv::Mat_<double>(4, 4) << 
        std::cos(angle), 0, std::sin(angle), tx,
        0, 1, 0, ty,
        -std::sin(angle), 0, std::cos(angle), tz,
        0, 0, 0, 1);
}

cv::Mat Map::rtz(double angle, double tx, double ty, double tz)
{
    return (cv::Mat_<double>(4, 4) << 
        std::cos(angle), -std::sin(angle), 0, tx,
        std::sin(angle), std::cos(angle), 0, ty,
        0, 0, 1, tz,
        0, 0, 0, 1);
}

cv::Mat Map::rt_vacuum()
{
    return rtz(CV_PI / 2.0, 50, 70, 0);
}

std::tuple<double, double, double> Map::get_robot_coordinates()
{
    Pose3d pose = pose_getter_();
    
    double y = -23.53 * (-31.95 - pose.y);
    double x = -23.58 * (-20.36 - pose.x);

    return {x, y, pose.yaw};
}

std::tuple<double, double, double> Map::get_robot_coordinates_with_noise()
{
    Pose3d pose = noisy_pose_getter_();
    
    double y = -23.53 * (-31.95 - pose.y);
    double x = -23.58 * (-20.36 - pose.x);

    return {x, y, pose.yaw};
}

void Map::reset()
{
}